[Core] Refactor ChibiOS USB endpoints to be fully async

KarlK90 opened the pull request:
## Description

This PR aims to unifies the ChibiOS USB endpoint implementation for all IN and OUT endpoints to fully async operation.

### Current ChibiOS USB endpoint implementation

Previously unidirectional sending for all IN endpoints was [implemented in a blocking manner](https://github.com/qmk/qmk_firmware/blob/ace510ffa0f8d51e1393ae9fcd17da00a5404c37/tmk_core/protocol/chibios/usb_main.c#L862-L881). Sending was implemented by invoking the ChibiOS USB API directly, waiting with a fixed timeout of 100ms when the endpoint wasn't ready and discarding any report that couldn't be send.

Bidirectional IN and OUT endpoints used a driver that was derived from the [ChibiOS CDC driver](https://github.com/qmk/qmk_firmware/pull/2750) that was already async in operation. This implementation wasn't suitable for unidirectional endpoints though and discarded any new report if the sending queue was full.

### Why asynchronous report handling instead of simple blocking?

The basic assumption is that we don't know when an IN endpoint is polled the next time or at all and we can't apply a heuristic for how long to wait. Likewise we don't know if an OUT endpoint will ever receive data. Some examples to back this up:

* The console endpoint is not regularly polled on Windows and Linux if no application e.g. `qmk console` is listening to this endpoint. See #5631.

* Simple USB stack implementations that are found in BIOSes might not support all QMK features like extrakey, CDC or MIDI endpoints but the user might still attempt to send out a USB report of this kind.

* After resuming from suspend a host might immediately poll an IN endpoint but the 2nd poll has a gap of multiple seconds. This was happening on Windows hosts see this AVR example #19871.

Furthermore USB HID keybord reports (and other absolute mode reports) are stateful. A pressed key report must be followed by a report that releases the key again or it will remain stuck.

From this I derive the following requirements:

1. We must not block the main processing loop of QMK when sending or receiving reports, otherwise we dead lock the device. See https://github.com/qmk/qmk_firmware/pull/22386#discussion_r1380026253
2. We must not discard the latest HID reports or we can bring the keyboard state out of sync. e.g. stuck keys.

Which leads to my conclusion that:

1. That sending and receiving must be implemented async
2. USB IN reports must be queued in a ring buffer which always holds the latest reports

Which is done in this PR.

As the USB idle rate and get report request handling was tightly coupled to the blocking sending implementation it was refactored as well to not introduce any regression in features. A benefit is that these are now supported for all IN endpoints in a generic manner.

## Tests

All tests are run against `handwired/onekey/<board>:default` keyboards with enabled `console`, `nkro` and `mousekeys` features. So the most used feature set is covered.

### Automated USB compliance tests

For compliance tests the [USB3CV](https://www.usb.org/document-library/usb3cv) tool version 3.0.0.0 is used on a W11 hosts. 

| MCU | Chapter 9 Tests (USB 2 Devices) | HID Tests |
| - | - | - |
| STM32F401 | :heavy_check_mark: | :heavy_check_mark: |
| STM32F411 | :heavy_check_mark: | :heavy_check_mark: |
| STM32F303 | :heavy_check_mark: | :heavy_check_mark: |
| RP2040 | :heavy_check_mark: | :heavy_check_mark: |

### Manual suspend and resume tests

These manual tests exercised that resume and suspend is working as expected, which means:

* Host is woken on keypress
* No hanging keys on wake
* Keyboard works just the same as before suspending

| MCU | Linux | MacOS | W11 |
| - | - | - | - |
| STM32F401 | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| STM32F411 | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| STM32F303 | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| RP2040 | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Inconsistent USB driver for ChibiOS based keyboards
* Edge cases for suspend after resume bugs with W11/W10
* Fix suspend and resume handling of QMK on ChibiOS powered MCUs which relied on hard reseting the USB stack as there where numerous unaddressed edge cases such as:
  * STM32F4s not waking the host
  * STM32F4s being unresponsive after wake-up
  * STM32F3s only sending the key press event but not sending the key up event after wake-up on W11 hosts. Leading to hanging keys.
* Properly implement the following requests for **all** endpoints:
  * set/get idle
  * get report

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

dexter93 commented:
Does this supersede #21537?

KarlK90 commented:
> Does this supersede #21537?

This supersedes 6a88a07da6ca15539a88f3181d6ee9de1b51b1b4 of #21537. The other fixes still apply.

drashna reviewed (approved):
aside from already dealt with mouse endpoint issues, works well in testing.

KarlK90 commented:
> Something about this seems to cause reports to get stuck.  Can reproduce on both moonlander (stm32f303) and blackpill based board.  

Which kind of reports? All of them or only keyboard reports? How does this manifest and is there a way to reproduce?

KarlK90 commented:
@drashna I can confirm that mouse scrolling was broken with this implementation as usb reports that where below the size of an endpoint would be buffered and sent in an incomplete manner. I've committed a fix. If you have the time please try it again.

drashna commented:
Yeah, this seems to work properly, and doesn't cause the issue with keys getting stuck.

KarlK90 commented:
Thanks for the confirmation! I'll wait for another review before merging though as the changes are rather profound and maybe there is an edge case that neither of us triggered during testing.

drashna commented:
unfortunately, the behavior seems to be back.  I'm not sure what is causing it specifically.   The define above seems to allow it to work a bit longer before going crazy.

And specifically, it looks like the last report sent gets stuck (eg, not cleared properly).  A key or two will work fine before the issue triggers.

KarlK90 commented:
You are using an STM32F4 board as the testing device right?

drashna commented:
STM32F411 and STM32F303

drashna commented:
> > STM32F411 and STM32F303
> 
> I've been using my F303 split keyboard for multiple hours and couldn't trigger the described behavior (mashing buttons and all that) on a host with a Linux OS (Ubuntu 23.04). Could you describe me some specific actions (e.g. oneshot mods combined with button mashing) that triggers those hangs?

honestly, it's frustratingly inconsistent.  right now, I can get it to trigger consistently on my moonlander (stm32f303), but no longer will occur.
I'm using arm-none-eabi-gcc 12.2.1 (from homebrew, on a mac), if it helps.

And I just tap a normal keycodes a couple of times and it triggers. Basic keycodes, too. 

And can confirm that it happens on RP2040, (adafruit macropad).

KarlK90 commented:
I really can't reproduce it on either W11 nor Linux. I'll grab a Mac from a friend and see if that triggers it.

drashna commented:
> I really can't reproduce it on either W11 nor Linux. I'll grab a Mac from a friend and see if that triggers it.

Well, if it makes you feel any better, on my W11 system, it doesn't produce the issue.  but same firmware on the board will on my mac.   No karabiner or the like on the system.

KarlK90 commented:
@drashna I borrowed a Mac from a friend and could reproduce the issue just fine. The bug is addressed in the latest commits. Let me know if this fixes your issues :-).

drashna commented:
Yeah, can confirm that the latest changes appear to fix the issues that I was seeing.

drashna commented:
To update this... it looks like when resuming from suspending, the entire board is unusable.  It stops sending *anything*.  This is with and without #21537 merged in, as well.

KarlK90 commented:
@drashna I've received my Mac Mini and with the latest changes my test keyboards work just fine after resuming from suspend. If you could test your setup that would be great, but I'm away for the next week so there is no rush. Thank you.

drashna commented:
Seems to work fine with F303 (moonlander), but still has issues with F411 (blackpill, tractyl manuform)

KarlK90 commented:
@drashna I currently lack a good stm32f411 setup to test the changes, could you give the latest interation one more spin?

drashna commented:
> @drashna I currently lack a good stm32f411 setup to test the changes, could you give the latest interation one more spin?

Sorry for the delay.  And looks like the stm32f405 is also effected (if that helps)

drashna commented:
Just tested this again, just in case, and yeah F411 and F405 don't respond at all, 12-oct-23

drashna commented on tmk_core/protocol/chibios/usb_main.c:
```diff
@@ -63,46 +46,28 @@ extern keymap_config_t keymap_config;
 #    define usb_lld_disconnect_bus(usbp)
 #endif
 
-uint8_t                keyboard_idle __attribute__((aligned(2)))     = 0;
```
Just a heads up, quantum/console.c  refers to `keyboard_idle` if PROTOCOL_VUSB isn't defined. Might be worth changing it

KarlK90 replied:
The idle handling got a big overhaul in this PR as well. For a short term fix I've deactivated the printing of the idle rate for now.

drashna commented on tmk_core/protocol/chibios/usb_main.c:
```diff
@@ -692,96 +310,60 @@ static bool usb_request_hook_cb(USBDriver *usbp) {
-    usb_request_hook_cb,   /* Requests hook callback */
-    usb_sof_cb             /* Start Of Frame callback */
+    usb_requests_hook_cb,  /* Requests hook callback */
+#if STM32_USB_USE_OTG1 == TRUE
```
I don't know how common it is, but would it be worth checking for otg2, as well? 
Mostly, just noticed that a lot of devices had it, when trying to figure out where the issue was popping up for me.

KarlK90 replied:
Thanks, added it to the check. Not sure if it is affected as well, as it is the High Speed capable iteration of the device, but won't hurt either.

drashna replied:
From the brief skimming, it should support FS, but requires the extra stuff for HS.  (but I could be wrong).

tzarc commented:
Marked for '24 Q1, will get it in early for testing.

KarlK90 commented:
> Marked for '24 Q1, will get it in early for testing.

Great, I'll add some more rationale and USB compliance tests to prove the overall design doesn't break anything.

KarlK90 commented:
I've updated the PR description with rationale why I think this PR is necessary and testing results. The commit messages have been updated as well, this is now ready for review.

KarlK90 commented:
Rebased to include #22560 which fixes the failing USB3CV compliance tests... which means that all QMK powered RP2040 keyboards are now fully USB spec compliant.

drashna commented:
Unfortunately, with this, the console doesn't seem to work properly.  Mac and Windows, and RP2040 and STM32F405 tested. 

It doesn't print anything unless it's printing a huge amount.

KarlK90 commented:
> Unfortunately, with this, the console doesn't seem to work properly. Mac and Windows, and RP2040 and STM32F405 tested.
> 
> It doesn't print anything unless it's printing a huge amount.

Can you post me a minimal reproducible example? I've used the following snippet:

```c
void housekeeping_task_kb(void) {
    static uint16_t now;

    if (timer_elapsed(now) > 10) {
        now = timer_read();
        printf("hello");
    }
}
```

Which won't print anything on `develop` and this pr because `qmk console` expects a newline at the end of any console output https://github.com/qmk/qmk_cli/blob/579763620d11463ad460ee9a31f7c03441984a0d/qmk_cli/subcommands/console.py#L130 otherwise it just discards the read bytes (which is questionable imho). `hid_listen` on the other hand works just fine as it just dumps everything it received from the device.

fauxpark commented:
> otherwise it just discards the read bytes (which is questionable imho)

Definitely not what should be happening. The code in the Toolbox, for example, accumulates received data and only actually outputs something when it hits a newline.

KarlK90 commented:
> > otherwise it just discards the read bytes (which is questionable imho)
> 
> Definitely not what should be happening. The code in the Toolbox, for example, accumulates received data and only actually outputs something when it hits a newline.

That would be the behavior that I would expect as well.

KarlK90 commented:
@drashna I stand corrected, Windows indeed didn't print anything as I have just tested. Linux is just more forgiving when it comes to shorter HID packages than they are described in the report descriptor... The solution was to always pad the console endpoints packets to full 32 bytes even if the string to be send is shorter than that. If you have the time please retry the latest state of this PR.

sigprof commented on tmk_core/protocol/chibios/usb_driver.c:
```diff
@@ -154,312 +78,254 @@ static void obnotify(io_buffers_queue_t *bqp) {
 }
 
-/** @} */
+void usb_endpoint_in_flush(usb_endpoint_in_t *endpoint, bool padded) {
+    osalDbgCheck(endpoint != NULL);
+
+    size_t count;
+    if (padded && obqGetFullBufferI(&endpoint->obqueue, &count) != NULL) {
+        ptrdiff_t padding_len = endpoint->config.buffer_size - count;
+        while (padding_len > 0) {
+            obqPutTimeout(&endpoint->obqueue, 0, TIME_IMMEDIATE);
+            padding_len--;
+        }
+    }
+    obqFlush(&endpoint->obqueue);
+}
```
This implementation does not look correct — `obqGetFullBufferI()` takes the buffer from the reader end of the queue, not the writer end.

Looks like ChibiOS does not have any API that would be suitable here, so you may try accessing `->ptr` and `->top` directly to determine how many bytes of padding are needed.

Alternatively, you may leave the buffers partially filled here, and pad the reports in the code which actually transmits USB packets (the buffer returned by `obqGetFullBufferI()` actually has the full size allocated, so you may fill the padding there).

KarlK90 replied:
Good catch! Thanks. I've updated to code to calculate the required padding bytes using `ptr` and `top`.

drashna commented:
New changes fix the issues with the console for me!

KarlK90 commented:
I've rebased onto `develop` and removed the `console` OUT endpoint which was done #22304 as well. The ring buffering implementation using `rbuf` is obsolete as all endpoints have an queue attached by default.

sigprof commented on tmk_core/protocol/chibios/usb_report_handling.h:
```diff
@@ -0,0 +1,77 @@
+
+#define QMK_USB_REPORT_STORAGE(_get_report, _set_report, _reset_report, _get_idle, _set_idle, _idle_timer_elasped, _report_count, _reports...) \
+    &((usb_report_storage_t){                                                                                                                  \
+        .reports            = (_Alignas(4) usb_fs_report_t *[_report_count]){_reports},                                                        \
```
It turned out that this kind of usage of `_Alignas` in compound literals is supported [only in GCC >= 8.x](https://gcc.gnu.org/legacy-ml/gcc-patches/2017-12/msg00406.html).  Because of that, GNU Guix is no longer suitable for building the QMK firmware (they have only `gcc-arm-none-eabi-7-2018-q2-update`).

This particular usage of `_Alignas` is also questionable — apparently it is applied to an array of pointers, which should be aligned appropriately anyway.  Other places where the alignment is applied to byte arrays used for endpoint buffers look reasonable though.
